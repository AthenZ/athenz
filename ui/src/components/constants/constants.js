/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as data from '../../config/zms.json';

export const MODAL_TIME_OUT = 2000;
export const GROUP_NAME_REGEX = data.types[3].StringTypeDef.pattern;
export const GROUP_MEMBER_NAME_REGEX = data.types[14].StringTypeDef.pattern;
export const GROUP_ROLES_CATEGORY = 'group-roles';
export const GROUP_MEMBER_PLACEHOLDER = 'user.<userid> or <domain>.<service>';
export const DISPLAY_SPACE = '\u23b5';

export const SERVICE_TYPE_DYNAMIC = 'dynamic';
export const SERVICE_TYPE_STATIC = 'static';
export const SERVICE_TYPE_DYNAMIC_LABEL = 'Dynamic';
export const SERVICE_TYPE_STATIC_LABEL = 'Static';
export const SEGMENTATION_TYPE_OUTBOUND = 'outbound';
export const SEGMENTATION_TYPE_INBOUND = 'inbound';
export const SEGMENTATION_TYPE_OUTBOUND_LABEL = 'Outbound';
export const SEGMENTATION_TYPE_INBOUND_LABEL = 'Inbound';
export const TOTAL_DYNAMIC_INSTANCES_DESC =
    'List of all dynamic instances(Both Active and Dormant).';
export const TOTAL_STATIC_INSTANCES_DESC = 'List of all static instances.';
export const TOTAL_HEALTHY_DYNAMIC_INSTANCES_LABEL =
    'Total Healthy Dynamic Instances';
export const TOTAL_STATIC_INSTANCES_LABEL = 'Total Static Instances';
export const TOTAL_HEALTHY_DYNAMIC_INSTANCES_DESC =
    'List of only active dynamic instances(i.e. instances refreshed certs within last 7 days).';
export const TOTAL_DYNAMIC_INSTANCES_LABEL = 'Total Dynamic Instances';

export const SERVICE_TABS = [
    {
        label: SERVICE_TYPE_STATIC_LABEL,
        name: SERVICE_TYPE_STATIC,
    },
    {
        label: SERVICE_TYPE_DYNAMIC_LABEL,
        name: SERVICE_TYPE_DYNAMIC,
    },
];

export const SEGMENTATION_CATEGORIES = [
    {
        label: SEGMENTATION_TYPE_INBOUND_LABEL,
        name: SEGMENTATION_TYPE_INBOUND,
    },
    {
        label: SEGMENTATION_TYPE_OUTBOUND_LABEL,
        name: SEGMENTATION_TYPE_OUTBOUND,
    },
];

export const SEGMENTATION_PROTOCOL_TYPE_TCP = 'TCP';
export const SEGMENTATION_PROTOCOL_TYPE_UDP = 'UDP';
export const SEGMENTATION_PROTOCOL_TYPE_ICMP = 'ICMP';
export const SEGMENTATION_PROTOCOL_TYPE_ICMPV6 = 'ICMPV6';
export const SEGMENTATION_PROTOCOL_TYPE_AH = 'AH';
export const SEGMENTATION_PROTOCOL_TYPE_ESP = 'ESP';
export const SEGMENTATION_PROTOCOL_TYPE_GRE = 'GRE';
export const SEGMENTATION_PROTOCOL_TYPE_IPV6_FRAGMENT = 'IPv6 Fragment';

export const SEGMENTATION_PROTOCOL = [
    {
        name: SEGMENTATION_PROTOCOL_TYPE_TCP,
        value: SEGMENTATION_PROTOCOL_TYPE_TCP,
    },
    {
        name: SEGMENTATION_PROTOCOL_TYPE_UDP,
        value: SEGMENTATION_PROTOCOL_TYPE_UDP,
    },
];

export const INSTANCE_RESOURCE_TYPE_IP_ADDRESS = 'IP Address';

export const STATIC_INSTANCES_RESOURCE_TYPES = [
    {
        name: INSTANCE_RESOURCE_TYPE_IP_ADDRESS,
        value: INSTANCE_RESOURCE_TYPE_IP_ADDRESS,
    },
];

export const DELETE_AUDIT_REFERENCE = 'deleted using Athenz UI';
