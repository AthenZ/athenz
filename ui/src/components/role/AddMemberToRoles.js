/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import AddModal from '../modal/AddModal';
import FlatPicker from '../flatpicker/FlatPicker';
import { colors } from '../denali/styles';
import Input from '../denali/Input';
import InputLabel from '../denali/InputLabel';
import styled from '@emotion/styled';
import Checkbox from '../denali/CheckBox';
import DateUtils from '../utils/DateUtils';
import NameUtils from '../utils/NameUtils';
import RequestUtils from '../utils/RequestUtils';

const SectionsDiv = styled.div`
    width: 800px;
    text-align: left;
    background-color: ${colors.white};
`;

const SectionDiv = styled.div`
    align-items: flex-start;
    display: flex;
    flex-flow: row nowrap;
    padding: 10px 30px;
`;

const ContentDiv = styled.div`
    flex: 1 1;
    display: flex;
    flex-flow: row wrap;
`;

const StyledInputLabel = styled(InputLabel)`
    flex: 0 0 150px;
    font-weight: 600;
    line-height: 36px;
`;

const StyledInput = styled(Input)`
    max-width: 500px;
    margin-right: 10px;
    width: 300px;
`;

const FlatPickrInputDiv = styled.div`
    margin-right: 10px;
    max-width: 500px;
    width: 260px;
    & > div input {
        position: relative;
        font: 300 14px HelveticaNeue-Reg, Helvetica, Arial, sans-serif;
        background-color: rgba(53, 112, 244, 0.05);
        box-shadow: none;
        color: rgb(48, 48, 48);
        height: 16px;
        min-width: 50px;
        text-align: left;
        border-width: 2px;
        border-style: solid;
        border-color: transparent;
        border-image: initial;
        border-radius: 2px;
        flex: 1 0 auto;
        margin: 0;
        margin-right: 10px;
        outline: none;
        padding: 0.6em 12px;
        transition: background-color 0.2s ease-in-out 0s,
            color 0.2s ease-in-out 0s, border 0.2s ease-in-out 0s;
        width: 80%;
    }
`;

const StyledRoleContainer = styled.div`
    width: 100%;
`;

const StyledRoles = styled.div`
    border-top: 1px solid #d8dade;
    height: 60%;
`;

const StyledRole = styled.div`
    background-color: rgba(53, 112, 244, 0.06);
    padding: 10px;
    width: calc(100% - 10px);
`;

const StyledJustification = styled(Input)`
    width: 300px;
    margin-top: 5px;
`;

export default class AddMemberToRoles extends React.Component {
    constructor(props) {
        super(props);
        this.api = this.props.api;
        this.onSubmit = this.onSubmit.bind(this);
        this.state = {
            showModal: !!this.props.showAddMemberToRoles,
            checkedRoles: [],
        };
        this.dateUtils = new DateUtils();
    }

    onSubmit() {
        if (!this.state.memberName || this.state.memberName === '') {
            this.setState({
                errorMessage: 'Member name is required.',
            });
            return;
        }

        if (
            this.props.justificationRequired &&
            (this.state.justification === undefined ||
                this.state.justification.trim() === '')
        ) {
            this.setState({
                errorMessage: 'Justification is required to add a member.',
            });
            return;
        }

        let member = {
            memberName: this.state.memberName,
            expiration:
                this.state.memberExpiry && this.state.memberExpiry.length > 0
                    ? this.dateUtils.uxDatetimeToRDLTimestamp(
                          this.state.memberExpiry
                      )
                    : '',
        };
        // send api call and then reload existing members component
        this.api
            .addMemberToRoles(
                this.props.domain,
                this.state.checkedRoles,
                this.state.memberName,
                member,
                this.state.justification
                    ? this.state.justification
                    : 'added using Athenz UI',
                this.props._csrf
            )
            .then(() => {
                this.setState({
                    showModal: false,
                    justification: '',
                });
                const successRoles = this.state.checkedRoles.join(', ');
                this.props.onSubmit(
                    `Successfully added ${this.state.memberName} to roles: ${successRoles}`
                );
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    inputChanged(key, evt) {
        this.setState({ [key]: evt.target.value });
    }

    onCheckboxChanged(role, event) {
        let checkedRoles = this.state.checkedRoles;
        if (checkedRoles.includes(role) && !event.target.checked) {
            checkedRoles = checkedRoles.filter((checkedRole) => {
                return checkedRole !== role;
            });
        } else {
            checkedRoles.push(role);
        }
        this.setState({ checkedRoles });
    }

    render() {
        let roleCheckboxes = [];
        this.props.roles.forEach((role) => {
            const roleName = NameUtils.getShortName(':role.', role.name);
            let onCheckboxChanged = this.onCheckboxChanged.bind(this, roleName);
            roleCheckboxes.push(
                <StyledRole key={roleName}>
                    <Checkbox
                        checked={this.state.checkedRoles.includes(roleName)}
                        label={roleName}
                        name={roleName}
                        onChange={onCheckboxChanged}
                        key={roleName}
                    />
                </StyledRole>
            );
        });

        let sections = (
            <SectionsDiv
                autoComplete={'off'}
                data-testid='add-member-to-roles-form'
            >
                <SectionDiv>
                    <StyledInputLabel htmlFor='member-name'>
                        Member
                    </StyledInputLabel>
                    <ContentDiv>
                        <StyledInput
                            id='member-name'
                            name='member-name'
                            value={this.state.memberName}
                            onChange={this.inputChanged.bind(
                                this,
                                'memberName'
                            )}
                            placeholder='user.<shortid> or <domain>.<service>'
                        />
                        <FlatPickrInputDiv>
                            <FlatPicker
                                onChange={(memberExpiry) => {
                                    this.setState({ memberExpiry });
                                }}
                                id='addMemberToRoles'
                                clear={this.state.memberExpiry}
                            />
                        </FlatPickrInputDiv>
                        {this.props.justificationRequired && (
                            <StyledJustification
                                id='justification'
                                name='justification'
                                value={this.state.justification}
                                onChange={this.inputChanged.bind(
                                    this,
                                    'justification'
                                )}
                                autoComplete={'off'}
                                placeholder='Enter justification here'
                            />
                        )}
                    </ContentDiv>
                </SectionDiv>
                <SectionDiv>
                    <StyledInputLabel htmlFor=''>Roles</StyledInputLabel>
                    <ContentDiv>
                        <StyledRoleContainer>
                            <StyledRoles>{roleCheckboxes}</StyledRoles>
                        </StyledRoleContainer>
                    </ContentDiv>
                </SectionDiv>
            </SectionsDiv>
        );

        return (
            <AddModal
                isOpen={this.state.showModal}
                cancel={this.props.onCancel}
                submit={this.onSubmit}
                title={'Add Member to Roles:'}
                errorMessage={this.state.errorMessage}
                sections={sections}
            />
        );
    }
}
