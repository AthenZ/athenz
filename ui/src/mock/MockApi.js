const origApi = require('../api').default;

export default class MockApi {
    static setMockApi(mockApi) {
        require('../api').default = () => {
            return mockApi;
        };
    }
    static cleanMockApi() {
        require('../api').default = () => {
            return origApi();
        };
    }
}
