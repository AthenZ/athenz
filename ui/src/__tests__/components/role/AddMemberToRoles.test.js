/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import AddMemberToRoles from '../../../components/role/AddMemberToRoles';
import { act } from 'react-dom/test-utils';
import {
    getStateWithUserList,
    renderWithRedux,
} from '../../../tests_utils/ComponentsTestUtils';
import MockApi from '../../../mock/MockApi';
import { USER_DOMAIN } from '../../../components/constants/constants';

describe('AddMemberToRoles', () => {
    afterEach(() => {
        MockApi.cleanMockApi();
    });

    it('should render', () => {
        let domain = 'domain';
        let roles = [];
        let role1 = {
            name: 'domain:role.role1',
            roleMembers: [
                { memberName: 'user.test1' },
                { memberName: 'user.test2' },
            ],
            memberExpiryDays: 30,
            serviceExpiryDays: 20,
            memberReviewDays: 70,
            serviceReviewDays: 80,
        };
        let role2 = {
            name: 'domain:role.role2',
            roleMembers: [
                { memberName: 'user.test3' },
                { memberName: 'user.test4' },
            ],
            memberExpiryDays: null,
            serviceExpiryDays: 20,
            memberReviewDays: 30,
            serviceReviewDays: null,
        };
        let roleMembers = [];
        let roleMember1 = {
            memberName: 'user.test1',
            memberRoles: [
                {
                    roleName: 'role1',
                },
            ],
            memberFullName: null,
        };
        roleMembers.push(roleMember1);
        let roleMember2 = {
            memberName: 'user.test2',
            memberRoles: [
                {
                    roleName: 'role1',
                },
            ],
            memberFullName: null,
        };
        roleMembers.push(roleMember2);
        let roleMember3 = {
            memberName: 'user.test3',
            memberRoles: [
                {
                    roleName: 'role2',
                },
            ],
            memberFullName: null,
        };
        roleMembers.push(roleMember3);
        let roleMember4 = {
            memberName: 'user.test4',
            memberRoles: [
                {
                    roleName: 'role2',
                },
            ],
            memberFullName: null,
        };
        roleMembers.push(roleMember4);

        MockApi.setMockApi({
            getRoles: jest.fn().mockReturnValue(Promise.resolve(roles)),
            getRoleMembers: jest
                .fn()
                .mockReturnValue(Promise.resolve(roleMembers)),
        });
        const onCancelMock = jest.fn();
        roles.push(role1);
        roles.push(role2);

        const { getByTestId } = renderWithRedux(
            <AddMemberToRoles
                domain={domain}
                // roles={roles}
                justificationRequired={true}
                onCancel={onCancelMock}
            />
        );
        const roleMemberForm = getByTestId('add-member-to-roles-form');
        expect(roleMemberForm).toMatchSnapshot();
    });

    it('search member to role', async () => {
        let domain = 'domain';
        let userList = { userList: [{ login: 'mock', name: 'Mock User' }] };
        const onCancelMock = jest.fn();
        renderWithRedux(
            <AddMemberToRoles
                domain={domain}
                justificationRequired={true}
                onCancel={onCancelMock}
                showAddMemberToRoles={true}
            />,
            getStateWithUserList(userList)
        );

        // change input to mocked user
        await act(async () => {
            fireEvent.change(
                screen.getByPlaceholderText(
                    `${USER_DOMAIN}.<shortid> or <domain>.<service>`
                ),
                {
                    target: { value: 'mock' },
                }
            );
        });

        // verify the correct input 'Mock User [user.mock]' is presented
        await waitFor(() =>
            expect(
                screen.getByText('Mock User [user.mock]')
            ).toBeInTheDocument()
        );
    });
});
