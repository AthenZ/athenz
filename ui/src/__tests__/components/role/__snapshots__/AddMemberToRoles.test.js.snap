// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMemberToRoles should render 1`] = `
<div
  data-testid="add-member-to-roles-form"
/>
`;
